presto-tablestore/src/main/java/com/facebook/presto/tablestore/IndexFirstSet.java

```diff
@@ -0,0 +1,251 @@
+        }
+        Optional<String> x = tables.stream().map(SchemaTableName::toString).reduce((a, b) -> a + "," + b);
+        return "[" + x.orElse("") + "]";
+    }
+
+    @SuppressWarnings("OptionalUsedAsFieldOrParameterType")
+    public static IndexFirstSet parseSearchIndexFirst(Optional<String> currentConnectionSchema, String hintKey, String hintValue)
+    {
+        if (StringUtils.isBlank(hintValue)) {
+            return IndexFirstSet.none();
+        }
+        String hint = hintValue.trim().toLowerCase(Locale.ENGLISH);
+        if (Type.auto.name().equals(hint)) {
+            return IndexFirstSet.auto();
+        }
+        if (Type.none.name().equals(hint)) {
+            return IndexFirstSet.none();
+        }
+        String pattern = "^" + Type.threshold + ":\\s*(\\d+)\\s*(%)?$";
+        if (hint.matches(pattern)) {
+            String err = "Invalid 'threshold' hint value '" + hintValue + "' of hint key '" + hintKey + "', regular pattern='" + pattern + "'";
+            Matcher m = Pattern.compile(pattern).matcher(hint);
+            checkState(m.find(), err);
+
+            int integer = Integer.parseInt(m.group(1));
+            String percent = m.group(2);
+
+            if (percent != null) {
+                if (integer < 1 || integer > MAX_VALUE_OF_INDEX_FIRST_MAX_PERCENT) {
+                    err = "Invalid 'threshold:${maxPercent}%' hint value '" + hintValue + "' of hint key '" + hintKey + "', "
+                            + "which[" + integer + "] should be within the range [1, " + MAX_VALUE_OF_INDEX_FIRST_MAX_PERCENT + "]";
+                    throw new IllegalArgumentException(err);
+                }
+                return thresholdWithPercent(integer);
+            }
+            else {
+                if (integer < 1 || integer > MAX_VALUE_OF_INDEX_FIRST_MAX_ROWS) {
+                    err = "Invalid 'threshold:${maxRows}' hint value '" + hintValue + "' of hint key '" + hintKey + "', "
+                            + "which[" + integer + "] should be within the range [1, " + MAX_VALUE_OF_INDEX_FIRST_MAX_ROWS + "]";
+                    throw new IllegalArgumentException(err);
+                }
+                return thresholdWithRows(integer);
+            }
+        }
+        if (hint.matches("^\\[[^\\[\\]]*]$")) {
+            hint = hint.substring(1, hint.length() - 1).trim();
+            if (StringUtils.isBlank(hint)) {
+                return IndexFirstSet.none();
+            }
+            String[] ss = hint.split("[ ;,]+");
+            Set<SchemaTableName> tables = Arrays.stream(ss)
+                    .filter(StringUtils::isNotBlank)
+                    .map(String::trim)
+                    .map(String::toLowerCase)
+                    .map(x -> checkAndAssemblySchemaTableName(currentConnectionSchema, hintKey, x))
+                    .collect(Collectors.toSet());
+            return IndexFirstSet.custom(tables);
+        }
+        String str = "use hint like these: " +
```

Comment:
thinking if we could extract 'str' as an constant 
can we have a meaningful name rather than 'str'?